### ISSUE TITLE
Some questions about train codes.

### ISSUE THREAD

**Issue by Kodapsy** [Jun-CEN/Open_world_3D_semantic_segmentation#24]
Hi, Jun-Cen.
I have been reading your code recently and have some questions about the tarin_incremental section to ask you.
First, The code in line 148 is to set the label 5 to 21 as a new class to learn incrementally, and then the following judgment is that if this batch of data does not contain the label 21, skip the next round, and there is no need to learn incrementally. Am I getting this right?
https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/semantickitti_scripts/train_cylinder_asym_ood_incremental.py#L148-L158
Second, I generally understand that this code is the processing of output_normal_dummy and voxel_label_origin before loss_normal is calculated, and the code in line 169 indicates that loss_normal is calculated with labels 0-19. But I have some doubts about the other codes, and just beginners can't understand exactly what they mean. If you have the time can you explain them line by line briefly? If you don't have time, simply explain the general meaning of each paragraph. Thank you very much.
https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/semantickitti_scripts/train_cylinder_asym_ood_incremental.py#L163-L175
Third, I can see that this code is some processing of output and label before calculating loss_dummy. I'm a little confused about line 180, why set the label 20 to 0? As with the second question, If you don't have time, simply explain the general meaning of each paragraph.  I'm more confused about this code than the previous one, and would appreciate it if you could explain it line by line. Thank you very much.
https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/semantickitti_scripts/train_cylinder_asym_ood_incremental.py#L176-L185
Last but not least, What exactly does line 675 mean? y_out_dummy is one dimension more than y_eval, and that dimension represents the prediction for the unknown class, right? Does the lack of dimension in y_eval mean that we don't need to validate unknown classes when we validate them?
https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/network/segmentator_3d_asymm_spconv.py#L675-L679
And then corresponding to the above is this line of validation code, here to change the label 20 to 5, the incremental learning set is 21, because the above y_eval less one dimension?
https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/semantickitti_scripts/train_cylinder_asym_ood_incremental.py#L114
I would appreciate it if you could take the time to answer. Wish you a happy life!

**Comment by Jun-CEN** [comment]
Hi,

First question, Yes. During incremental learning, we only use the data that contains the new classes.

Second question. dis_label_origin means the pseudo labels of old classes, which is from here: https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/dataloader/pc_dataset.py#L199
It means that we use the old model to predict the pseudo labels for old classes because we assume that the labels of old classes are not available during incremental learning. The ground truth of the whole point cloud is composed of the ground truth of the new class and pseudo labels of the old classes.

Third question. The label of 20 represents our pseudo unknown objects. https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/dataloader/dataset_semantickitti.py#L646
This paragraph means the loss that each point should has the second largest probability of the unknown class, which is the L_cal^uk in the paper. So for the pseudo unknown objects we don't use this loss. This is why we set them to 0 to ignore in the loss.

Last question. The first dimension of of y_out means the probability of the newly learned class. The rest of y_out means the probability of the unknown class. So y_out_dummy means 0 (ignore class) + 19 (known class) + 1 (unknown class) + 1 (newly learned class). During training, we use y_out_dummy, so the 21 represents newly learned class. During inference, y_eval is  0 (ignore class) + 19 (known class) + 1 (newly learned class), so that we can evaluate both old and new classes. The inference of the unknown class is calculated by the AUROC not IoU, so there is no unknown class in y_eval. In such a case, 20 means the newly learned class.

**Comment by Kodapsy** [comment]
> Hi,
> 
> First question, Yes. During incremental learning, we only use the data that contains the new classes.
> 
> Second question. dis_label_origin means the pseudo labels of old classes, which is from here:
> 
> https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/dataloader/pc_dataset.py#L199
> 
> 
> It means that we use the old model to predict the pseudo labels for old classes because we assume that the labels of old classes are not available during incremental learning. The ground truth of the whole point cloud is composed of the ground truth of the new class and pseudo labels of the old classes.
> Third question. The label of 20 represents our pseudo unknown objects.
> 
> https://github.com/Jun-CEN/Open_world_3D_semantic_segmentation/blob/084c341291ae186f174743d8fae5efecea89122f/dataloader/dataset_semantickitti.py#L646
> 
> 
> This paragraph means the loss that each point should has the second largest probability of the unknown class, which is the L_cal^uk in the paper. So for the pseudo unknown objects we don't use this loss. This is why we set them to 0 to ignore in the loss.
> Last question. The first dimension of of y_out means the probability of the newly learned class. The rest of y_out means the probability of the unknown class. So y_out_dummy means 0 (ignore class) + 19 (known class) + 1 (unknown class) + 1 (newly learned class). During training, we use y_out_dummy, so the 21 represents newly learned class. During inference, y_eval is 0 (ignore class) + 19 (known class) + 1 (newly learned class), so that we can evaluate both old and new classes. The inference of the unknown class is calculated by the AUROC not IoU, so there is no unknown class in y_eval. In such a case, 20 means the newly learned class.

Thank you for your patience，wish you a happy life.
